File: packages/alera_browser/windows/browser_page_methods.cpp
#include "alera_browser_plugin.h"

#include "browser_page.h"
#include "browser_profile.h"
#include "browser_value.h"

#include <algorithm>
#include <cstring>
#include <map>
#include <memory>
#include <optional>

namespace alera_browser {
namespace {

std::shared_ptr<BrowserPage> RequiredPage(
    AleraBrowserPlugin* plugin,
    const EncodableMap& arguments,
    MethodResultPtr* result) {
  const auto id = StringValue(arguments, "pageId");
  const auto page = id.has_value() ? plugin->FindPage(*id) : nullptr;
  if (!page) {
    Error(
        std::move(*result), "page_not_found",
        "The browser page does not exist.");
  }
  return page;
}

bool SafeHeaderName(const std::string& value) {
  if (value.empty()) {
    return false;
  }
  constexpr char kSeparators[] = "()<>@,;:\\\"/[]?={} \t";
  return std::all_of(value.begin(), value.end(), [&](unsigned char item) {
    return item > 0x20 && item < 0x7f &&
           std::strchr(kSeparators, item) == nullptr;
  });
}

bool SafeHeaderValue(const std::string& value) {
  return std::none_of(value.begin(), value.end(), [](unsigned char item) {
    return (item < 0x20 && item != '\t') || item == 0x7f;
  });
}

std::optional<std::map<std::string, std::string>> Headers(
    const EncodableMap& arguments) {
  std::map<std::string, std::string> result;
  for (const auto& [key, value] : MapValue(arguments, "headers")) {
    const auto* name = std::get_if<std::string>(&key);
    const auto* text = std::get_if<std::string>(&value);
    if (name == nullptr || text == nullptr ||
        !SafeHeaderName(*name) || !SafeHeaderValue(*text)) {
      return std::nullopt;
    }
    result[*name] = *text;
  }
  return result;
}

}  // namespace

bool HandleBrowserPageMethod(
    AleraBrowserPlugin* plugin,
    const std::string& method,
    const EncodableMap& arguments,
    MethodResultPtr result) {
  if (method == "page.create") {
    const auto profile_id =
        StringValue(arguments, "profileId").value_or("default");
    const auto profile = plugin->FindProfile(profile_id);
    if (!profile) {
      Error(
          std::move(result), "profile_not_found",
          "The browser profile does not exist.");
      return true;
    }
    auto id = StringValue(arguments, "id");
    if (!id.has_value()) {
      id = "page-" + std::to_string(plugin->next_page_id_++);
    }
    if (!IsValidBrowserId(*id) || plugin->FindPage(*id)) {
      Error(
          std::move(result), "invalid_page",
          "The page id is invalid or already exists.");
      return true;
    }
    const auto opener_id = StringValue(arguments, "openerPageId");
    if (opener_id.has_value()) {
      const auto opener = plugin->FindPage(*opener_id);
      if (!opener || opener->profile_id() != profile_id) {
        Error(
            std::move(result), "invalid_opener",
            "The popup opener must exist in the same profile.");
        return true;
      }
    }
    const auto initial_url = StringValue(arguments, "initialUrl");
    if (initial_url.has_value() && !IsAllowedBrowserUrl(*initial_url)) {
      Error(
          std::move(result), "invalid_url",
          "Only HTTP, HTTPS, and about:blank are allowed.");
      return true;
    }
    auto shared_result =
        std::shared_ptr<MethodResult>(std::move(result));
    const auto lifetime = plugin->lifetime();
    BrowserPage::Create(
        plugin, plugin->parent_window(), profile, *id, initial_url,
        StringValue(arguments, "userAgent"), opener_id,
        BoolValue(arguments, "transient"),
        [plugin, lifetime, shared_result](
            HRESULT created,
            std::shared_ptr<BrowserPage> page) {
          if (lifetime.expired()) {
            if (page) {
              page->Close();
            }
            return;
          }
          if (FAILED(created) || !page) {
            shared_result->Error(
                "page_create_failed", HResultMessage(created));
            return;
          }
          if (!plugin->AddPage(page)) {
            page->Close();
            shared_result->Error(
                "duplicate_page", "The browser page already exists.");
            return;
          }
          shared_result->Success(EncodableValue(EncodableMap{
              {EncodableValue("id"), EncodableValue(page->id())},
              {EncodableValue("title"), EncodableValue("")}}));
        });
    return true;
  }

  const auto id = StringValue(arguments, "pageId");
  if (method == "page.close") {
    if (id.has_value()) {
      plugin->RemovePage(*id);
    }
    Success(std::move(result));
    return true;
  }
  auto page = RequiredPage(plugin, arguments, &result);
  if (!page) {
    return true;
  }

  if (method == "page.attach" || method == "page.detach") {
    page->SetAttached(method == "page.attach");
    Success(std::move(result));
  } else if (method == "page.setObscured") {
    page->SetObscured(BoolValue(arguments, "obscured"));
    Success(std::move(result));
  } else if (method == "page.setBounds") {
    page->SetBounds(
        DoubleValue(arguments, "x"), DoubleValue(arguments, "y"),
        DoubleValue(arguments, "width"),
        DoubleValue(arguments, "height"),
        DoubleValue(arguments, "scale", 1));
    Success(std::move(result));
  } else if (method == "page.adoptTransient") {
    const auto profile_id = StringValue(arguments, "profileId");
    if (!page->transient() || !profile_id.has_value() ||
        page->profile_id() != *profile_id) {
      Error(
          std::move(result), "invalid_transient_page",
          "The transient popup does not match the requested profile.");
    } else {
      page->SetAdopted(true);
      Success(std::move(result));
    }
  } else if (method == "page.promoteTransient") {
    if (!page->transient() || !page->adopted()) {
      Error(
          std::move(result), "invalid_transient_page",
          "The transient popup must be adopted before promotion.");
    } else {
      page->Promote();
      Success(std::move(result));
    }
  } else if (method == "page.currentUrl") {
    const auto value = page->CurrentUrl();
    Success(
        std::move(result),
        value.has_value() ? EncodableValue(*value) : EncodableValue());
  } else if (method == "page.title") {
    const auto value = page->Title();
    Success(
        std::move(result),
        value.has_value() ? EncodableValue(*value) : EncodableValue());
  } else if (method == "page.canGoBack") {
    Success(std::move(result), EncodableValue(page->CanGoBack()));
  } else if (method == "page.canGoForward") {
    Success(std::move(result), EncodableValue(page->CanGoForward()));
  } else if (
      method == "page.goBack" || method == "page.goForward" ||
      method == "page.reload" || method == "page.stop") {
    const HRESULT operation =
        method == "page.goBack"       ? page->GoBack()
        : method == "page.goForward" ? page->GoForward()
        : method == "page.reload"    ? page->Reload()
                                      : page->Stop();
    if (FAILED(operation)) {
      Error(
          std::move(result), "navigation_failed",
          HResultMessage(operation));
    } else {
      Success(std::move(result));
    }
  } else if (method == "page.loadUrl") {
    const auto url = StringValue(arguments, "url");
    const auto headers = Headers(arguments);
    if (!url.has_value() || !IsAllowedBrowserUrl(*url)) {
      Error(
          std::move(result), "invalid_url",
          "Only HTTP, HTTPS, and about:blank are allowed.");
      return true;
    }
    if (!headers.has_value()) {
      Error(
          std::move(result), "invalid_headers",
          "Navigation headers contain invalid characters.");
      return true;
    }
    auto shared_result =
        std::shared_ptr<MethodResult>(std::move(result));
    page->Navigate(*url, *headers, [shared_result](HRESULT value) {
      if (FAILED(value)) {
        shared_result->Error("navigation_failed", HResultMessage(value));
      } else {
        shared_result->Success();
      }
    });
  } else if (method == "page.evaluateJavaScript") {
    const auto script = StringValue(arguments, "script");
    if (!script.has_value()) {
      Error(
          std::move(result), "invalid_script",
          "A JavaScript expression is required.");
      return true;
    }
    auto shared_result =
        std::shared_ptr<MethodResult>(std::move(result));
    page->ExecuteScript(
        *script, [shared_result](HRESULT value, std::string json) {
          if (FAILED(value)) {
            shared_result->Error(
                "javascript_failed", HResultMessage(value));
          } else {
            shared_result->Success(EncodableValue(std::move(json)));
          }
        });
  } else if (method == "page.upload") {
    Error(
        std::move(result), "unsupported_upload",
        "Native file upload is not advertised by this backend.");
  } else {
    result->NotImplemented();
    return false;
  }
  return true;
}

}  // namespace alera_browser

File: packages/alera_browser/macos/Classes/BrowserPageMethods.swift
import FlutterMacOS
import WebKit

extension AleraBrowserPlugin {
  func handlePageMethod(
    _ method: String,
    arguments: [String: Any],
    result: @escaping FlutterResult
  ) throws {
    if method == "page.create" {
      try createPage(arguments: arguments, result: result)
      return
    }
    let pageID = try arguments.requiredString("pageId")
    guard let page = pages[pageID] else {
      if method == "page.close" {
        result(nil)
        return
      }
      throw BrowserMethodError("page_not_found", "The browser page does not exist.")
    }
    switch method {
    case "page.attach":
      try page.attach()
      result(nil)
    case "page.detach":
      page.detach()
      result(nil)
    case "page.setObscured":
      page.obscured = arguments.bool("obscured")
      page.updateSurfaceVisibility()
      result(nil)
    case "page.setBounds":
      try page.setBounds(
        x: arguments.double("x"),
        y: arguments.double("y"),
        width: arguments.double("width"),
        height: arguments.double("height"),
        scale: arguments.double("scale", default: 1)
      )
      result(nil)
    case "page.adoptTransient":
      let profileID = try arguments.requiredString("profileId")
      guard page.transient, page.profile.id == profileID else {
        throw BrowserMethodError(
          "invalid_transient_page",
          "The transient popup does not match the requested profile."
        )
      }
      page.adopted = true
      result(nil)
    case "page.promoteTransient":
      guard !page.transient || page.adopted else {
        throw BrowserMethodError(
          "transient_not_adopted",
          "The transient popup must be adopted before it is promoted."
        )
      }
      page.transient = false
      result(nil)
    case "page.close":
      closePage(pageID)
      result(nil)
    case "page.loadUrl":
      try loadURL(arguments: arguments, page: page)
      result(nil)
    case "page.currentUrl":
      result(page.webView.url?.absoluteString)
    case "page.title":
      result(page.webView.title)
    case "page.canGoBack":
      result(page.webView.canGoBack)
    case "page.canGoForward":
      result(page.webView.canGoForward)
    case "page.goBack":
      page.webView.goBack()
      result(nil)
    case "page.goForward":
      page.webView.goForward()
      result(nil)
    case "page.reload":
      page.webView.reload()
      result(nil)
    case "page.stop":
      page.webView.stopLoading()
      result(nil)
    case "page.evaluateJavaScript":
      try evaluateJavaScript(arguments: arguments, page: page, result: result)
    case "page.upload":
      try uploadFiles(arguments: arguments, page: page, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func createPage(arguments: [String: Any], result: @escaping FlutterResult) throws {
    let id = arguments.optionalString("id") ?? UUID().uuidString.lowercased()
    try validateIdentifier(id, kind: "page", maximum: 128)
    guard pages[id] == nil else {
      throw BrowserMethodError("duplicate_page", "The browser page already exists.")
    }
    let profileID = arguments.optionalString("profileId") ?? "default"
    let selectedProfile = try profile(profileID)
    let openerID = arguments.optionalString("openerPageId")
    if let openerID {
      guard let opener = pages[openerID], opener.profile.id == profileID else {
        throw BrowserMethodError(
          "invalid_opener",
          "The opener page must exist in the same browser profile."
        )
      }
    }
    let page = BrowserPage(
      owner: self,
      id: id,
      profile: selectedProfile,
      openerPageID: openerID,
      transient: arguments.bool("transient")
    )
    if let userAgent = arguments.optionalString("userAgent") {
      page.webView.customUserAgent = userAgent
    }
    pages[id] = page
    if let initialURL = arguments.optionalString("initialUrl") {
      do {
        try page.webView.load(URLRequest(url: validatedBrowserURL(initialURL)))
      } catch {
        pages.removeValue(forKey: id)
        page.close()
        throw error
      }
    }
    var value: [String: Any] = ["id": id]
    if let title = page.webView.title { value["title"] = title }
    result(value)
  }

  private func loadURL(arguments: [String: Any], page: BrowserPage) throws {
    let url = try validatedBrowserURL(arguments.requiredString("url"))
    var request = URLRequest(url: url)
    if let headers = arguments["headers"] as? [String: String] {
      for (name, value) in headers {
        guard
          !name.isEmpty,
          !name.contains("\r"),
          !name.contains("\n"),
          !value.contains("\r"),
          !value.contains("\n")
        else {
          throw BrowserMethodError("invalid_header", "A browser request header is invalid.")
        }
        request.setValue(value, forHTTPHeaderField: name)
      }
    }
    page.webView.load(request)
  }

  private func evaluateJavaScript(
    arguments: [String: Any],
    page: BrowserPage,
    result: @escaping FlutterResult
  ) throws {
    let script = try arguments.requiredString("script")
    page.webView.evaluateJavaScript(script) { value, error in
      if let error {
        result(BrowserMethodError("javascript_failed", error.localizedDescription).asFlutterError)
      } else {
        result(value)
      }
    }
  }

  private func uploadFiles(
    arguments: [String: Any],
    page: BrowserPage,
    result: @escaping FlutterResult
  ) throws {
    guard page.pendingUploadResult == nil else {
      throw BrowserMethodError("upload_in_progress", "A file upload is already in progress.")
    }
    let elementRef = try arguments.requiredString("elementRef")
    guard
      let paths = arguments["filePaths"] as? [String],
      !paths.isEmpty
    else {
      throw BrowserMethodError("invalid_upload", "At least one file path is required.")
    }
    let urls = try paths.map { path -> URL in
      guard NSString(string: path).isAbsolutePath else {
        throw BrowserMethodError("invalid_upload", "Every upload path must be absolute.")
      }
      let url = URL(fileURLWithPath: path).standardizedFileURL
      var isDirectory: ObjCBool = false
      guard
        FileManager.default.fileExists(atPath: url.path, isDirectory: &isDirectory),
        !isDirectory.boolValue,
        FileManager.default.isReadableFile(atPath: url.path)
      else {
        throw BrowserMethodError("invalid_upload", "Every upload file must exist and be readable.")
      }
      return url
    }
    let reference = try javascriptLiteral(elementRef)
    let script = """
      (() => {
        const state = window.__aleraBrowserAutomation;
        const entry = state?.elements?.get(\(reference));
        if (!entry?.element?.isConnected) return JSON.stringify({ok:false,code:'stale_element'});
        if (JSON.stringify(state.signatureFor(entry.element)) !== JSON.stringify(entry.signature)) {
          return JSON.stringify({ok:false,code:'stale_element'});
        }
        if (!(entry.element instanceof HTMLInputElement) || entry.element.type !== 'file') {
          return JSON.stringify({ok:false,code:'wrong_element_type'});
        }
        entry.element.click();
        return JSON.stringify({ok:true});
      })()
      """
    page.pendingUploadURLs = urls
    page.pendingUploadResult = result
    page.webView.evaluateJavaScript(script) { [weak page] value, error in
      guard let page else { return }
      if let error {
        page.finishUpload(
          error: BrowserMethodError("upload_failed", error.localizedDescription)
        )
        return
      }
      guard
        let text = value as? String,
        let data = text.data(using: .utf8),
        let response = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
        response["ok"] as? Bool == true
      else {
        page.finishUpload(
          error: BrowserMethodError("upload_failed", "The file input is stale or invalid.")
        )
        return
      }
      DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak page] in
        guard let page, page.pendingUploadResult != nil else { return }
        page.finishUpload(
          error: BrowserMethodError(
            "upload_chooser_unavailable",
            "WebKit did not open the requested file input."
          )
        )
      }
    }
  }

  func closePage(_ pageID: String) {
    guard let page = pages.removeValue(forKey: pageID) else { return }
    cancelDecisions(forPage: pageID)
    cancelDownloads(forPage: pageID)
    page.close()
  }

  func page(_ id: String) throws -> BrowserPage {
    guard let page = pages[id], !page.closed else {
      throw BrowserMethodError("page_not_found", "The browser page does not exist.")
    }
    return page
  }
}

extension BrowserPage {
  func finishUpload(error: BrowserMethodError? = nil) {
    guard let result = pendingUploadResult else { return }
    pendingUploadResult = nil
    pendingUploadURLs = nil
    result(error?.asFlutterError)
  }
}

func validatedBrowserURL(_ value: String) throws -> URL {
  guard let url = URL(string: value), let scheme = url.scheme?.lowercased() else {
    throw BrowserMethodError("invalid_url", "The browser URL is invalid.")
  }
  let allowed = scheme == "http" || scheme == "https" || value == "about:blank"
  guard allowed, scheme == "about" || url.host != nil else {
    throw BrowserMethodError("unsupported_url", "Only HTTP, HTTPS, and about:blank are supported.")
  }
  return url
}
